Add PrecompiledToken

ThunderDeliverer opened the pull request:
Added PrecompiledToken example. The example introduces the precompiled
and predeployed smart contracts of the Acala EVM+ as well as the ADDRESS
utility of @acala-network/contracts dependency.

ThunderDeliverer committed:
Add PrecompiledToken

Added PrecompiledToken example. The example introduces the precompiled
and predeployed smart contracts of the Acala EVM+ as well as the ADDRESS
utility of @acala-network/contracts dependency.

ThunderDeliverer committed:
Add PrecompiledToken to run.sh

shunjizhan commented on precompiled-token/src/getACAinfo.ts:
```diff
@@ -0,0 +1,67 @@
+
+import { evmChai } from '@acala-network/bodhi';
+
+const Token = require('@acala-network/contracts/build/contracts/Token.json');
```
should we also use `import` to be more consistent?

ThunderDeliverer replied:
Sure, I don't have the preference and could find arguments for both (being concise vs. showing different options), It seems that you have a preference, so I changed this :ok_hand:

shunjizhan commented on precompiled-token/src/getACAinfo.ts:
```diff
@@ -0,0 +1,67 @@
+    console.log("Token total supply:", totalSupply.toString());
+    console.log("Token balance of %s is: %s", walletAddress, balance.toString());
+
+    const formattedSupply = balanceFormatting(totalSupply.toString(), decimals);
```
`ethers` lib has a util to do this, reference: https://github.com/AcalaNetwork/bodhi.js/blob/master/eth-rpc-adapter/src/__tests__/e2e/endpoint.test.ts#L506

ThunderDeliverer replied:
It was a different util than this one, but it was a good suggestion. Thank you! :raised_hands:

shunjizhan commented on precompiled-token/src/setup.ts:
```diff
@@ -0,0 +1,41 @@
+import {
```
can we merge this PR and import it from util directly? https://github.com/AcalaNetwork/waffle-tutorials/pull/2/files#diff-7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94R21

ThunderDeliverer replied:
Done :white_check_mark:

shunjizhan commented on precompiled-token/test/PrecompiledToken.test.ts:
```diff
@@ -0,0 +1,50 @@
+import { evmChai, Signer, TestProvider } from '@acala-network/bodhi';
+import { WsProvider } from '@polkadot/api';
+
+const Token = require('@acala-network/contracts/build/contracts/Token.json');
```
here also, should we also use `import`, so we won't mix cjs and esm

shunjizhan commented on precompiled-token/test/PrecompiledToken.test.ts:
```diff
@@ -0,0 +1,50 @@
+use(solidity);
+use(evmChai);
+
+const provider = new TestProvider({
```
here we can also reuse the `getTestProvider` from this PR, to be more DRY and so CI can change the URL consistently https://github.com/AcalaNetwork/waffle-tutorials/pull/2/files#diff-7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94R8

ThunderDeliverer committed:
Added setup utility to PrecompiledToken

Added setup utility to PrecompiledToken tutorial and refrined the
information script. README was u pdated to reflect these changes.

ThunderDeliverer commented:
@shunjizhan would you mind verifying that this tutorial doesn't conflict with our CI, as it has no `build` script?

shunjizhan commented:
Just checked it, and it is already compatible with the CI interface (so CI should be able to run these examples), but for the tests I still got a these `Some("NoPermission")` errors when running with `mandala-node:95c483d7`. Are they already passing with new mandala versions that don't have docker version yet?

I think we can either merge them, then fix tests, or wait for the tests to be fixed and merge them together. Basically if `./run.sh test` all passed then we are all good

ThunderDeliverer commented:
> Just checked it, and it is already compatible with the CI interface (so CI should be able to run these examples), but for the tests I still got a these `Some("NoPermission")` errors when running with `mandala-node:95c483d7`. Are they already passing with new mandala versions that don't have docker version yet?
> 
> I think we can either merge them, then fix tests, or wait for the tests to be fixed and merge them together. Basically if `./run.sh test` all passed then we are all good

The test pass for me with dockerized local network.. Are you using the latest version? I just pulled the repository to make sure my version of bodhi.js is up to date
